lib/Dialect/ESI/ESIOps.cpp

```diff
@@ -126,6 +128,48 @@ void WrapValidReady::build(OpBuilder &b, OperationState &state, Value data,
         b.getI1Type(), data, valid);
 }
 
+namespace {
+/// Eliminate back-to-back wrap-unwraps to reduce the number of ESI channels.
+struct RemoveWrapUnwrap : public OpConversionPattern<WrapValidReady> {
+public:
+  RemoveWrapUnwrap(MLIRContext *ctxt) : OpConversionPattern(ctxt) {}
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult match(Operation *op) const final {
+    auto wrap = dyn_cast<WrapValidReady>(op);
+    if (!wrap)
+      return failure();
+    if (!wrap.chanOutput().hasOneUse())
+      return failure();
+    if (!isa<UnwrapValidReady>(wrap.chanOutput().use_begin()->getOwner()))
+      return failure();
+    return success();
+  }
+
+  void rewrite(WrapValidReady wrap, ArrayRef<Value> operands,
+               ConversionPatternRewriter &rewriter) const final {
+    UnwrapValidReady unwrap =
+        dyn_cast<UnwrapValidReady>(wrap.chanOutput().use_begin()->getOwner());
+    assert(unwrap && "Must be operating on wrap-unwrap pair");
+
+    auto tmpOp = rewriter.create<mlir::ConstantOp>(
+        wrap.getLoc(), rewriter.getIntegerAttr(rewriter.getI1Type(), 0));
+
+    // Replacing `tmpOp` with nullptr also "works".
+    rewriter.replaceOp(wrap, {tmpOp, unwrap.ready()});
```

Comment:
Not sure about the difference between the pattern drivers, but I do think maybe I was misleading with my last comment.  I think `unwrap` does need to be replaced, not erased. But shouldn't `wrap` just be erased instead of replaced with a temp or nullptr?

Reply:
`replaceOp` is the only way exposed by `rewriter` to set the consumer of `esi.wrap`'s ready connection (backpressure).

Reply:
It has to be plumbed through from whomever is driving `unwrap`'s ready input.

Reply:
I see, that makes sense. I'm still not sure about the tmp/nullptr. Wouldn't you want to replace both return values of `wrap` with meaningful values? Something like `rewriter.replaceOp(wrap, {unwrap.chanInput(), unwrap.ready()})`?

Anyway, I'm just guessing here. I will checkout this branch and try it out for myself

Reply:
`unwrap.chanInput()` is driven by `wrap.chanOutput()`, so I'm concerned what happens to that value when `wrap` gets erased.

Thanks for checking it out. I could definitely use another set of eyes on this problem.